### PR TITLE
Serialization: Serializable meta annotation

### PR DESCRIPTION
**Pull request by rsinukov**
https://youtrack.jetbrains.com/issue/KTOR-2704

Add handling for (not yet released) kotlinx.serialization `@MetaSerializable`

### CODE REVIEW

**Review by Repoman124 — CHANGES_REQUESTED** [review]
#4583

**Review by Repoman124 — COMMENTED** [review]
I'm attempting my first pull

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/GeneratorHelpers.kt [hunk 1]
```diff
@@ -597,7 +604,9 @@ interface IrBuilderExtension {
                     irProperty.backingField?.symbol ?: error("Property expected to have backing field"),
                     property.type.toIrType(),
                     receiver
-                )
+                ).let {
+                    if (property.type.toIrType().isKClass()) kClassExprToJClassIfNeeded(startOffset, endOffset, it) else it
```
[the remark this answers is not in the document]
Comment (rsinukov) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/GeneratorHelpers.kt]: done

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/resolve/NamingConventions.kt [hunk 2]
```diff
@@ -38,6 +38,8 @@ object SerializationAnnotations {
     internal val contextualOnPropertyFqName = FqName("kotlinx.serialization.Contextual")
     internal val polymorphicFqName = FqName("kotlinx.serialization.Polymorphic")
     internal val additionalSerializersFqName = FqName("kotlinx.serialization.UseSerializers")
+
+    internal val metaSerializableAnnotationFqName = FqName("kotlinx.serialization.MetaSerializable")
```
[the remark this answers is not in the document]
Comment (rsinukov) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/resolve/NamingConventions.kt]: done

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt [hunk 3]
```diff
@@ -46,17 +46,7 @@ class SerializableIrGenerator(
...
-
-        for (i in 0 until annot.valueArgumentsCount) {
-            if (annot.getValueArgument(i) != null) return false
-        }
-
-        return true
-    }
-
-    private val IrClass.isInternalSerializable: Boolean get() = kind == ClassKind.CLASS && hasSerializableAnnotationWithoutArgs()
+    private val IrClass.isInternalSerializable: Boolean get() = this.descriptor.isInternalSerializable
```
[the remark this answers is not in the document]
Comment (rsinukov) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt]: done

File: plugins/kotlin-serialization/kotlin-serialization-compiler/testData/boxIr/metaSerializable.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+fun testCustomSerializerAndMetaAnnotation() {
+//    val string = Json.encodeToString(Project3.serializer(), Project3("name", "lang"))
+//    assertEquals("""name:lang""", string)
+//
+//    val reconstructed = Json.decodeFromString(Project3.serializer(), string)
+//    assertEquals("name", reconstructed.name)
+//    assertEquals("lang", reconstructed.language)
+//
+//    val info = Project3.serializer().descriptor.annotations.filterIsInstance<MySerializableWithInfo>().first()
```
[the remark this answers is not in the document]
Comment (rsinukov) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/testData/boxIr/metaSerializable.kt:117]: sorry, fotgot to delete it. done

File: plugins/kotlin-serialization/kotlin-serialization-compiler/testData/boxIr/metaSerializable.kt [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+class Project2(val name: String, val language: String)
+
+@Serializable
+class Wrapper(
+//    @MySerializableWithInfo(234, Int::class) val project: Project2
+)
+
+@Serializable(with = MySerializer::class)
+@MySerializableWithInfo(123, String::class)
+class Project3(val name: String, val language: String)
```
[the remark this answers is not in the document]
Comment (rsinukov) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/testData/boxIr/metaSerializable.kt:40]: Why only in runtime? 
Added tests to this repo.

**Review by sandwwraith — APPROVED** [review]

**Pull request merged**